source/Halibut/Transport/ActiveTcpConnectionsLimiter.cs

```diff
@@ -0,0 +1,115 @@
+﻿// Copyright 2012-2013 Octopus Deploy Pty. Ltd.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+using System;
+using System.Collections.Generic;
+using System.Runtime.CompilerServices;
+using Halibut.Diagnostics;
+using Halibut.Exceptions;
+
+namespace Halibut.Transport
+{
+    public interface IActiveTcpConnectionsLimiter
+    {
+        IDisposable ClaimAuthorizedTcpConnection(Uri subscriptionId);
+
+        IDisposable CreateUnlimitedLease();
+    }
+
+    public class ActiveTcpConnectionsLimiter : IActiveTcpConnectionsLimiter
+    {
+        readonly HalibutTimeoutsAndLimits timeoutsAndLimits;
+
+        Dictionary<Uri, StrongBox<int>> activeConnectionCountPerSubscriptionId = new();
+
+        public ActiveTcpConnectionsLimiter(HalibutTimeoutsAndLimits timeoutsAndLimits)
+        {
+            this.timeoutsAndLimits = timeoutsAndLimits;
+        }
+
+        public IDisposable ClaimAuthorizedTcpConnection(Uri subscriptionId)
+        {
+            //if there is no limit, then we return a NoOp lease (which doesn't limit anything)
+            if (!timeoutsAndLimits.MaximumActiveTcpConnectionsPerPollingSubscription.HasValue)
+            {
+                return CreateUnlimitedLease();
+            }
+
+            return new LimitingAuthorizedTcpConnectionLease(subscriptionId, activeConnectionCountPerSubscriptionId, timeoutsAndLimits.MaximumActiveTcpConnectionsPerPollingSubscription.Value);
+        }
+
+        public IDisposable CreateUnlimitedLease()
+        {
+            return new UnlimitedAuthorizedTcpConnectionLease();
+        }
+
+        class UnlimitedAuthorizedTcpConnectionLease : IDisposable
+        {
+            public void Dispose()
+            {
+            }
+        }
+
+        class LimitingAuthorizedTcpConnectionLease : IDisposable
+        {
+            readonly Uri subscriptionId;
+            readonly Dictionary<Uri, StrongBox<int>> activeConnectionCountPerSubscriptionId;
+
+            public LimitingAuthorizedTcpConnectionLease(Uri subscriptionId, Dictionary<Uri, StrongBox<int>> activeConnectionCountPerSubscriptionId, int maximumAcceptedTcpConnectionsPerThumbprint)
+            {
+                this.subscriptionId = subscriptionId;
+                this.activeConnectionCountPerSubscriptionId = activeConnectionCountPerSubscriptionId;
+
+                lock (this.activeConnectionCountPerSubscriptionId)
```

Comment:
Nit: the activeConnectionCountPerSubscriptionId is both a dictionary and a lock

I have wondered if a type like this would make that more clear.
```
class LimitConcurrentAccessToObject<T> {
  private T t; // set by ctor
  private object mutex = new object();
  
  public void WithExclusiveAccess(Action<T> action) {
      lock(mutex) {
          action(t);
      }
  }
}
```

This means the code would turn into:

```
f(LimitConcurrentAccessToObject<Dictionary<Uri, StrongBox<int>>> exclusiveDict) {
   exclusiveDict.WithExclusiveAccess(dict => {
      // code that works on the dict in a way you almost certainly can not get wrong.
   });
}
```

Just a thought :) which you are free to do nothing with

Reply:
I like the dictionary as a lock because it kinda implies that you are locking access around the dictionary